### PR TITLE
ci: enforce workflow hygiene invariants

### LINKED ISSUE #2413: Workflow lint gate: concurrency, artifact-action majors, --init everywhere, pinned SHA consumed downstream
**Priority 3 of 6 in #2411 — after items 1–2.**

Part of #2411 (plan item 3). Kills the #2385/#2388/#2391/#2396 class mechanically.

## Objective

Extend `tests/test_issue2231_workflow_hygiene.py` so the following are asserted over every workflow and every container-spawning script, each with a planted-offence test proving the scanner can fail:

1. `concurrency:` present on every dispatchable (`workflow_dispatch` / `push` / `schedule` / `release`) workflow (#2391).
2. `actions/upload-artifact` and `actions/download-artifact` majors match across producer and consumer jobs of the same workflow chain (#2385).
3. `--init` on every container invocation, including `docker run` in `scripts/` and `tests/smoke/**` (#2396 — the guard that pinned this was retired; the successor never carried the assertion forward; `test_container_jobs_pass_init` covers only workflow `container:` blocks today).
4. A SHA pinned in a `prepare`-style job is the SHA consumed downstream — no re-reading live refs after a pin (#2387, #2406).

## Required reading

- `tests/test_issue2231_workflow_hygiene.py` (existing scanners: env-map, runner series, egress, actionlint, container `--init`).
- #2385, #2388, #2391, #2396, #2387, #2406 bodies and comments.
- `.github/workflows/*.yml`, `scripts/local-smoke*.sh`, any `docker run` under `scripts/` and `tests/smoke/`.
- `.agents/policy/testing.md`.

## Constraints

- Lint lands green on `devel`: either the offending workflows are fixed first (blocked-by relations wired on this ticket) or the fixes ride in the same PR with their own red-first proof — never a skip-list of known offenders.
- Every new assertion needs a planted-offence test (scanner catches a synthetic violation) — no vacuous negative assertions.
- No changes under `src/`.

## Verification

- WHEN each new scanner runs against a fixture with the violation planted THEN it fails naming the file and offence; WHEN run on `devel` after the fixes THEN green.
- WHEN a workflow gains `workflow_dispatch` without `concurrency:` THEN pytest fails.
- WHEN a `docker run` without `--init` is added under `scripts/` or `tests/smoke/` THEN pytest fails.
- `scripts/agent/run-gates.sh --diff <base>` green.

## Coverage matrix

Grep at ticket start: `on:` triggers per workflow, every `upload-artifact@`/`download-artifact@`, every `docker run`/`container:`, every `git clone`/`checkout` `ref:` fed from a prior job's output. Table in the PR body.

## Escalation

If rule 4 (pinned SHA consumed downstream) cannot be expressed statically for a workflow, STOP and checkpoint with the case; do not weaken the rule to a substring check.

### PR DESCRIPTION
**Pull request by andrebrait**
## Summary

- require top-level concurrency on every externally dispatchable workflow, preserving lossless operational queues
- enforce artifact-action major parity through needs, reusable-call instances, and `workflow_run` display-name chains
- enforce exact pre-image `--init` on workflow containers and tracked shell/Python/PHP Docker invocations
- enforce exact prepare/read/resolve SHA consumption without derived refs, shadowed aliases, or live re-resolution

Fixes #2413

## Source-derived coverage matrix

| Rule | Source rows | Result |
|---|---|---|
| External-trigger concurrency | 24 `.yml`/`.yaml` workflow documents; 14 missing blocks fixed, 10 existing blocks preserved | clean; image-refresh and nightly-failure-alert use `queue: max` |
| Artifact majors | 26 action rows and 14 local reusable-call jobs: build-pkg-linux → image/smoke/UI; smoke-single → smoke aggregate/module-durations; release attach/external consumers; independent Nightly chain | affected producer/consumer chains aligned on v8; Nightly remains v7 end-to-end |
| Container init | every workflow job `container` mapping plus every tracked UTF-8 file under `scripts/**` and `tests/smoke/**` | no current production invocation; preventive scanner clean |
| Pinned SHA consumption | Nightly prepare source/ports/tools/matrix; Release prepare/read-matrix source/CI metadata/ports; release-published resolve source | all nine pins reach exact dependent ref/identity sinks; no derived/live replacement |

Concurrency fixes cover `agent-config`, `build-image`, `build-pkg-linux`, `coderabbit-cli`, `context-budget`, `image-refresh`, `nightly-failure-alert`, `repo-install`, `smoke-single`, `smoke`, `test-version-matrix`, `top1m-healthcheck`, `ui-tests`, and `version-tracker`.

## Hostile/planted coverage

The scanners exercise empty/invalid inventories and YAML; quoted and mixed triggers/action refs; inline comments; `.yml`/`.yaml`; missing/cyclic needs and reusable workflows; duplicate display names/artifacts; unresolved outputs; per-call reusable inputs; unmatched `workflow_run` selectors; semver and unclassifiable artifact refs in uncalled reusable-only workflows; dynamic names/globs; block-local container options; Docker continuations, heredocs, wrappers, long and short global/run value options, image boundary, option terminators, module-, function-, and factory-local Python argv/import resolution, PHP/shell variables, and late `--init`; exact SHA fallbacks plus fake outputs, format/decorated expressions, derived/fallback job env, step-env shadowing, untrusted alternatives, exact sinks beside invalid aliases, semicolon-delimited live resolution, multiline live resolution, and `readonly` identity aliases.

## Frozen RED → GREEN

The original scanner blob `5e4d1f944b24636cbfbf139871bf4e3ff948d7c0` was proven against untouched base `8016210c99e602e6718fed068232ae4aef7fbaa2`: `4 failed, 24 passed`; every hostile scanner test passed and failures were solely real base workflow offences. The same blob passed `28 passed` after workflow reconciliation.

The earlier final adversarial oracle blob `ff07e098e05fef314a7a8fc63bd7efd9fe8fbf70` produced `8 failed in 0.07s` against exact pre-fix scanner `3e0ae17f688628e25f11b3c377b48876b2b125dc`, then `32 passed in 0.57s` unchanged on the candidate.

### Final bypass closure

Frozen oracle: `tests/test_issue2245_stateless_nightly.py` blob `e3ff77873ccd69f502c0f68ba60a666b5b8a7f68`.

RED used that exact oracle against the scanner bytes from pre-fix head `942942e50ca8c9f38ac13833dedc4cdebe7a1d4c` via isolated module import:

```text
uv run python -c 'import pytest, subprocess, types; source=subprocess.check_output(["git","show","HEAD:tests/test_issue2231_workflow_hygiene.py"], text=True); pre=types.ModuleType("pre_hygiene"); pre.__file__="tests/test_issue2231_workflow_hygiene.py"; exec(compile(source, pre.__file__, "exec"), pre.__dict__); P=type("P", (), {"pytest_runtest_setup": lambda self, item: setattr(item.module, "hygiene", pre)}); nodes=["tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_derived_env_alias_beside_exact_checkout", "tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_reports_semicolon_live_replacement_beside_exact_sink", "tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_resolves_function_local_subprocess_import"]; raise SystemExit(pytest.main(["-q", "-n0", *nodes], plugins=[P()]))'
```

```text
3 failed in 0.06s
```

The three REDs were exact checkout beside derived/untrusted `SOURCE_SHA`, exact `git show` beside semicolon-delimited live `FRESH`, and function-local `import subprocess` before `docker run` without `--init`.

GREEN on commit `46eb31b33a9e8665abca4b7ea13d67253d271f98`:

```text
uv run pytest -q tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_derived_env_alias_beside_exact_checkout tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_reports_semicolon_live_replacement_beside_exact_sink tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_resolves_function_local_subprocess_import
```

```text
3 passed in 0.56s
```

The oracle remained byte-identical at `e3ff77873ccd69f502c0f68ba60a666b5b8a7f68` through GREEN.

### Final bounded alias-sink closure

Alias-sink scanner at that closure: `tests/test_issue2231_workflow_hygiene.py` blob `e5501756118a4611cd42e96f07f98bccd516e6e0`.

Frozen oracle: `tests/test_issue2245_stateless_nightly.py` blob `143a31b187ce8761d509ed74597a8e237e73acbd`.

RED ran the three exact hostile rows against unchanged scanner head `5a59644ec4ea9a037ea81b15e294caa3f6860147`:

```text
uv run pytest -q tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_derived_env_alias_at_equals_flag_identity_sink tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_reports_live_alias_at_equals_flag_identity_sink tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_derived_shell_alias_at_flag_identity_sink
pytest: 3 failed in 0.61s
```

All three returned `[]`: derived/untrusted job env at `--ports-ref=$PORTS_SHA`, live `ls-remote` replacement at `--ports-ref=$FRESH`, and quoted parameter concatenation consumed through `--ports-ref "$FRESH"`.

GREEN ran the unchanged oracle after the generic flag/alias fix:

```text
uv run pytest -q tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_derived_env_alias_at_equals_flag_identity_sink tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_reports_live_alias_at_equals_flag_identity_sink tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_derived_shell_alias_at_flag_identity_sink
pytest: 3 passed in 0.62s
```

The oracle remained byte-identical at `143a31b187ce8761d509ed74597a8e237e73acbd` through GREEN. The scanner now parses both `--*-ref VALUE` / `--*-sha VALUE` and equals forms, and propagates exact, derived, overwritten, and live provenance through shell-local aliases without flag- or file-specific exemptions.
### Final unknown flag-sink closure

Current final scanner: `tests/test_issue2231_workflow_hygiene.py` blob `ce1d7728de632654f100c9b19903202a68d51ce5`.

Frozen oracle: `tests/test_issue2245_stateless_nightly.py` blob `9d58e41c5000b5b438693c176748b3ab36316189`.

RED ran the exact unknown-alias, split/equals direct-literal, and scope-control rows against unchanged scanner head `322bb4287978f1acbb9bf8e3e5e881c57c8ed58d`:

```text
uv run pytest -q -n0 tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_unknown_alias_at_flag_identity_sink tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_direct_literal_at_flag_identity_sink tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_scopes_flag_identity_sinks_to_matching_prepared_pins
pytest: 3 failed, 1 passed in 0.11s
```

The three hostile rows all returned `[]`: `FRESH=main` passed through split `--ports-ref "$FRESH"`, and direct `main` passed through both `--ports-ref main` and `--ports-ref=main`. The passing control proved exact prepared aliases/direct needs outputs remain valid and unrelated flags or workflows without a matching prepared pin remain out of scope.

GREEN on commit `3c1e4f22d9ae076921a421cec9d742e4e6f9d298`:

```text
uv run pytest -q -n0 tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_unknown_alias_at_flag_identity_sink tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_rejects_direct_literal_at_flag_identity_sink tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_scopes_flag_identity_sinks_to_matching_prepared_pins
pytest: 4 passed in 0.05s
```

The oracle remained byte-identical at `9d58e41c5000b5b438693c176748b3ab36316189` through GREEN. Matching `--*-ref` / `--*-sha` sinks now fail closed for unknown aliases and direct literals in split or equals form while retaining exact prepared aliases/direct outputs and ignoring unrelated flag families or workflows with no matching prepared pin.

## Final sibling-flag closure

Current final scanner: `tests/test_issue2231_workflow_hygiene.py` blob `17660024a15fa3253f752c04a210f673fc6a0593`.

Frozen oracle: `tests/test_issue2245_stateless_nightly.py` blob `14213f1c8beaccdc53078e27d11e5177736af7e3`.

RED ran the split and equals sibling-flag rows plus the exact-SHA/raw-ref control against unchanged scanner head `3c1e4f22d9ae076921a421cec9d742e4e6f9d298`:

```text
uv run pytest -q -n0 tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_validates_each_identity_flag_argument tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_scopes_flag_identity_sinks_to_matching_prepared_pins
pytest: 2 failed, 1 passed in 0.08s
```

Both hostile rows returned `[]`: an exact prepared `--ports-ref` masked a second `--ports-ref main` argument in split and equals form. The passing control proved an exact `--ports-sha` may still be paired with raw `--ports-ref` metadata.

GREEN on commit `71f89c4b6c38d75cb41e3d2f90e8e7a4f23920a3`:

```text
uv run pytest -q -n0 tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_validates_each_identity_flag_argument tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_scopes_flag_identity_sinks_to_matching_prepared_pins
pytest: 3 passed in 0.08s
```

The oracle remained byte-identical at `14213f1c8beaccdc53078e27d11e5177736af7e3` through GREEN. Each relevant identity argument is now validated independently; a verified same-base `--*-sha` keeps paired `--*-ref` metadata out of identity selection without a file-specific exemption.

## Final multi-assignment closure

Current final scanner: `tests/test_issue2231_workflow_hygiene.py` blob `6af015ff7dee7779597449573debb1618c359e98`.

Frozen oracle: `tests/test_issue2245_stateless_nightly.py` blob `51a1bf6ab83d572226b24d74e99b7792bceb2687`.

RED ran the exact `export OTHER=1 PORTS_SHA=main` and `readonly OTHER=1 PORTS_SHA=main` hostile rows plus unrelated-assignment controls against unchanged pre-fix scanner blob `17660024a15fa3253f752c04a210f673fc6a0593`:

```text
uv run pytest -q -n0 tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_tracks_every_exported_pin_alias tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_keeps_exact_pin_after_unrelated_exported_assignments
pytest: 2 failed, 2 passed in 0.11s
```

Both hostile rows returned `[]`: only `OTHER=1` was processed, so the later `PORTS_SHA=main` overwrite escaped the sibling `--ports-ref "$PORTS_SHA"` identity sink. Both controls passed with two unrelated assignments and the exact prepared `PORTS_SHA` unchanged.

GREEN on commit `c11a4b7e7358df8d368cb04f66a15f0bc03d15c3`:

```text
uv run pytest -q -n0 tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_tracks_every_exported_pin_alias tests/test_issue2245_stateless_nightly.py::test_workflow_hygiene_keeps_exact_pin_after_unrelated_exported_assignments
pytest: 4 passed in 0.07s
```

The oracle remained byte-identical at `51a1bf6ab83d572226b24d74e99b7792bceb2687` through GREEN. The scanner now processes every assignment in `export` and `readonly` commands in command order, preserving exact, derived, overwritten, and live provenance without variable- or file-specific exemptions.

## Focused gates

```text
uv run pytest -q -n0 tests/test_issue2231_workflow_hygiene.py tests/test_issue2245_stateless_nightly.py tests/test_nightly_workflow_contract.py tests/test_smoke_leg_zero_selection_gate.py
pytest: 115 passed in 2.64s
```

```text
uv run ruff check tests/test_issue2231_workflow_hygiene.py tests/test_issue2245_stateless_nightly.py
All checks passed!
uv run ruff format --check tests/test_issue2231_workflow_hygiene.py tests/test_issue2245_stateless_nightly.py
2 files already formatted
```

```text
uv run mypy tests/
Success: no issues found in 355 source files
```

Commit policy hooks passed. The final bounded change touches only `tests/test_issue2231_workflow_hygiene.py` and `tests/test_issue2245_stateless_nightly.py`; no workflow, `src/`, dependency, or documentation file changed.

Final pushed head: `c11a4b7e7358df8d368cb04f66a15f0bc03d15c3`.

Parent validation and landing remain pending at this exact head. This head is not a merge instruction.

## Final exact-byte scanner proof

Current final scanner: `tests/test_issue2231_workflow_hygiene.py` blob `6af015ff7dee7779597449573debb1618c359e98` at head `c11a4b7e7358df8d368cb04f66a15f0bc03d15c3`.

An independent detached-base run copied these exact committed scanner bytes onto untouched `8016210c99e602e6718fed068232ae4aef7fbaa2` and ran:

```text
uv run pytest -n 0 tests/test_issue2231_workflow_hygiene.py -q
pytest: 4 failed, 24 passed
```

Every hostile/synthetic row passed; the four failures were only the enumerated real base workflow offences. The same exact blob on the candidate produced:

```text
pytest: 28 passed
```

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

### CODE REVIEW

**Comment by andrebrait** [comment]
@coderabbitai review

**Comment by andrebrait** [comment]
## Contract-conformance audit — round 1

Reviewed head: `a84f58742cc118885ad8df32eabdf3a71a7e9c4c` (full PR diff).

Findings ledger:

1. `blocking` Python function-local Docker argv was outside the scanned binding scope — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554` with lexical/local/factory fixtures.
2. `blocking` multiline `git ls-remote` command substitution could feed a downstream pin sink — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554` with multiline and `readonly` provenance checks.

Every workflow/config production hunk otherwise conformed to issue #2413. Re-review will focus on the fixes from this reviewed SHA.

Role: OMP pfBlockerNG contract reviewer, read-only.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Correctness + hostile-input audit — round 1

Reviewed head: `a84f58742cc118885ad8df32eabdf3a71a7e9c4c` (full PR diff).

Findings ledger:

1. `blocking` Docker global options and dynamic shell/Python/PHP commands could bypass discovery — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554`.
2. `blocking` Docker run option arity could treat an option value as the image and reject valid pre-image `--init` — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554` with the run-option arity inventory.
3. `blocking` semver artifact action refs were ignored — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554`; valid semver majors parse and unknown refs fail closed.
4. `blocking` a pinned-SHA sink could fall back to a live expression — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554`; every alternative must be a known prepared pin.
5. `blocking` `readonly` alias overwrite/live re-resolution was missed — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554` with assignment provenance and exact git-sink diagnostics.

All 18 non-scanner changed files were considered and fine. Re-review will execute the recorded bypasses against the fix.

Role: OMP pfBlockerNG correctness reviewer, read-only.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Test-honesty audit — round 1

Reviewed head: `a84f58742cc118885ad8df32eabdf3a71a7e9c4c` (full PR diff).

Findings ledger:

1. `blocking` PR evidence named the obsolete `d460…` scanner blob — `fixed`: PR body now names final blob `5e4d1f944b24636cbfbf139871bf4e3ff948d7c0`, exact-base `4 failed, 24 passed`, and unchanged candidate `28 passed`.
2. `blocking` decorated/fallback/fake SHA sinks were under-tested and accepted — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554` with format, appended-expression, literal/untrusted fallback, and fake-output fixtures.
3. `blocking` empty workflow inventory could pass vacuously — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554` with a planted empty-root failure.

Review-regression assertions are separate from the scanner implementation in `tests/test_issue2245_stateless_nightly.py`; their RED probes failed on the pre-fix scanner and the final focused matrix is green.

Role: OMP pfBlockerNG test-honesty reviewer, read-only.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Over-engineering audit — round 1

Reviewed head: `a84f58742cc118885ad8df32eabdf3a71a7e9c4c` (full PR diff).

Findings ledger:

1. `nitpick` dead `_ShaPin.step_id` state — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554`.
2. `nitpick` hand-written start/end expression identity check — `fixed@5be3383212eea90aac09c5c6fadfa1d5f4e5a554` by reusing the existing expression parser for exact prepared-pin alternatives.

No workflow/config complexity finding. Net requested cut was four lines; the fixes remove the dead field and replace the broad identity predicate with the shared parser.

Role: OMP ponytail reviewer, read-only.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Contract-conformance audit — final

- Head reviewed: `c11a4b7e7358df8d368cb04f66a15f0bc03d15c3`
- Tier / effort: mid / medium
- Verdict: **CLEAN**

All original #2413 rows and every later hostile row are closed: external-trigger concurrency; chain-aware artifact-major parity; exact pre-image `--init`; exact prepared-pin sinks; reusable-only and `merge_group` roots; call-instance selectors; lexical Docker discovery; Git globals/live substitutions; separated/equals/duplicate identity flags; derived, live, unknown, literal, export, and readonly aliases. Final contract matrix: `97 passed`.

All 19 changed files were examined. No new finding remains.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Correctness + hostile-input audit — final

- Head reviewed: `c11a4b7e7358df8d368cb04f66a15f0bc03d15c3`
- Tier / effort: top / high
- Verdict: **CLEAN**

Executed bypass probes now emit the exact expected diagnostics. Closed classes include derived/fallback env aliases beside exact sinks, semicolon-delimited live `ls-remote`, lexical and function-local subprocess imports, duplicate separated/equals identity flags, derived/live/unknown/direct flag arguments, and multi-name `export`/`readonly` overwrites. Full prior hostile probe chain reports no bypass; concurrency collision controls remain false.

All workflow/config changes were considered and remain correct. No blocking correctness or security finding remains.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Test-honesty audit — final

- Head reviewed: `c11a4b7e7358df8d368cb04f66a15f0bc03d15c3`
- Tier / effort: small / high
- Verdict: **CLEAN**

Current blobs: scanner `6af015ff7dee7779597449573debb1618c359e98`; oracle `51a1bf6ab83d572226b24d74e99b7792bceb2687`; smoke `f3abb33040e9f25664e5a1f972cf078417906b12`.

The PR now records the final scanner byte proof: untouched-base `4 failed, 24 passed`, candidate `28 passed`. Layered frozen hostile proofs cover each subsequent scanner bypass, including the final multi-assignment oracle (`2 failed, 2 passed` → `4 passed`, byte-identical). Final focused matrix: `115 passed`; independent verifier matrix: `119 passed`. Mutation probes for duplicate producers, Docker diagnostics, concurrency losslessness, identity flags, and alias provenance all fail on regression.

No vacuous assertion or stale evidence remains.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Over-engineering audit — final

- Head reviewed: `c11a4b7e7358df8d368cb04f66a15f0bc03d15c3`
- Tier / effort: top / high
- Verdict: **NITPICKS ONLY; clean for merge**

The final scanner additions directly implement reproduced bypasses and reuse existing `shlex`, expression parsing, and provenance structures. No dead or speculative production mechanism was added. The prior redundant artifact pre-scan was removed.

One optional duplication remains: `download_selectors` can be collapsed for about two net lines. It is non-blocking and was left explicit for per-call diagnostic clarity. Net optional removal: 2 lines.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

File: tests/test_issue2231_workflow_hygiene.py [hunk 1]
```diff
@@ -309,3 +320,1557 @@ def test_pass_silencer_scanner_catches_every_spelling() -> None:
...
+                    available = (*upstream, *produced)
+                    matched = {
+                        artifact
+                        for selector in selectors
+                        for artifact in available
+                        if _selectors_match(artifact.name, selector)
+                    }
+                    if not matched:
+                        continue
+                    download_matches[key] = True
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/test_issue2231_workflow_hygiene.py]: APPLY — fixed in the landed review-fix chain. Reusable download state is keyed by call stack and resolved selectors; planted same-reusable good/bad caller fixtures fail on conflation and pass on the final scanner. The final artifact hostile matrix is green at `c11a4b7e7358df8d368cb04f66a15f0bc03d15c3`.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Review orchestrator — final ledger

Head: `c11a4b7e7358df8d368cb04f66a15f0bc03d15c3`

### Findings

1. Empty workflow discovery could pass vacuously — **fixed**; `_workflow_sources()` fails closed and planted empty-root coverage passes.
2. Reusable-workflow download state conflated call instances/selectors — **fixed**; exact caller-instance keys and hostile fixtures pass.
3. Uncalled reusable and `merge_group` artifact roots bypassed traversal — **fixed**.
4. Docker option terminators, global/run arity, dynamic wrappers, lexical/factory argv, and function-local imports bypassed discovery — **fixed** with exact diagnostics.
5. Prepared-pin sinks accepted decorated/fallback/fake outputs, shadowed aliases, Git globals/direct live substitutions, semicolon live assignments, derived/live/unknown/direct identity flags, duplicate sibling flags, and multi-name `export`/`readonly` overwrites — **fixed** with frozen hostile rows.
6. Artifact semver/unclassified refs, duplicate producers, and call-chain provenance gaps — **fixed**.
7. Stale scanner evidence — **fixed**; PR body records final scanner `6af015ff…`, untouched-base `4 failed, 24 passed`, candidate `28 passed`, plus every layered frozen oracle.
8. Optional duplicated selector map — **skipped**: two-line reduction is non-blocking and the explicit map preserves per-call diagnostics.

### Gates

- Independent verifier: **PASS**; final focused matrix `119 passed`; base/final RED→GREEN and hashes matched.
- Four adversarial legs: contract **CLEAN**, correctness **CLEAN**, test honesty **CLEAN**, over-engineering **NITPICKS ONLY**.
- CodeRabbit finished once on `a84f587`; both findings above were applied. Later changes are scanner/test review fixes, so no second review spend.
- Exact-head canonical gates: `GATES: PASS` (locked pytest, Ruff, format, mypy).
- Exact-head CI: `PASS`; required checks green. UI/benchmark jobs skipped by intended path scope.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.
